### PR TITLE
feat: Deferred Revenue and Expense report with actual and upcoming postings

### LINKED ISSUE #28225: Deferred Accounting Report for Income and Expenses
### Problem Statement:

1.  Since income and expenses are deferred, you don't come to know:
                   1.  The actual income booked in a current month
                   2.  The upcoming income which will be posted in coming months, as per start and end date
4.  The upcoming expense ------------------ // -----------------
5.  The expected against actual post
6.  Actual posting of expense as per the start and end date for each fiscal year / quarter / month
7.  Actual posting of income as per the start and end date for each fiscal year / quarter / month
 
### Expected Solution:

1.  Make new reports to provide this information

### Who is the end user? Define persona(s).

1.  Business owner, who wants to:
2.  see income and expense, with and without deferred posting
3.  Upcoming Income / Expense as per future deferred posting
4.  Account Manager / Auditor, who wants to review if the posting of deferred entries has happened correctly as per the Start and End Period set in the invoices.

### PR DESCRIPTION
**Pull request by ruthra-kumar**
### Deferred Revenue and Expense Report
This report aim to provide a simplified view of all deferred income/expense in a certain period. It can help in identifying the actual and expected posting.

### Deferred Revenue
def_revenue

### Deferred Expense
def_expense

### Documentation
https://docs.erpnext.com/docs/v13/user/manual/en/accounts/deferred_revenue/expense_report

### CODE REVIEW

File: erpnext/accounts/report/deferred_revenue_and_expense/deferred_revenue_and_expense.py [hunk 1]
```diff
@@ -0,0 +1,436 @@
...
+		elif items[0].deferred_expense_account:
+			self.type = "Purchase"
+
+		self.items = []
+		# for each uniq items
+		self.uniq_items = set([x.item for x in items])
+		for item in self.uniq_items:
+			self.items.append(Deferred_Item(item, self, [x for x in items if x.item == item]))
+
+	def get_postings(self):
```
Comment (deepeshgarg007) [inline erpnext/accounts/report/deferred_revenue_and_expense/deferred_revenue_and_expense.py]: Seems like a dead function, not called from anywhere

File: erpnext/accounts/report/deferred_revenue_and_expense/deferred_revenue_and_expense.js [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+			"options": [
+				{ "value": "Revenue", "label": __("Revenue") },
+				{ "value": "Expense", "label": __("Expense") }
+			],
+			"default": "Revenue",
+			"reqd": 1
+		},
+		{
+			"fieldname" : "with_upcoming_postings",
+			"label": __("Show with upcoming revenue/expense"),
```
Comment (deepeshgarg007) [inline erpnext/accounts/report/deferred_revenue_and_expense/deferred_revenue_and_expense.js:88]: This also shows unbooked values for the previous months, may better to call this label as "Show expected revenue/expense" and this should be unchecked by default

**Review by deepeshgarg007 — CHANGES_REQUESTED** [review]

**Comment by deepeshgarg007** [comment]
Fix the linting issues

File: erpnext/accounts/report/deferred_revenue_and_expense/deferred_revenue_and_expense.js (hunk 2 above)
Comment (ruthra-kumar) [reply erpnext/accounts/report/deferred_revenue_and_expense/deferred_revenue_and_expense.js:88]: Agreed. 
Along with this, I think it's better if it only shows the actual posting(bar graph) when unchecked. Thoughts?

Comment (deepeshgarg007) [reply erpnext/accounts/report/deferred_revenue_and_expense/deferred_revenue_and_expense.js:88]: If someone wants to compare the expected and actual posting, how do they do that then?

File: erpnext/accounts/report/deferred_revenue_and_expense/deferred_revenue_and_expense.py (hunk 1 above)
Comment (ruthra-kumar) [reply erpnext/accounts/report/deferred_revenue_and_expense/deferred_revenue_and_expense.py]: removed

File: erpnext/accounts/report/deferred_revenue_and_expense/deferred_revenue_and_expense.js (hunk 2 above)
Comment (ruthra-kumar) [reply erpnext/accounts/report/deferred_revenue_and_expense/deferred_revenue_and_expense.js:88]: renamed and 'unchecked' by default.

**Comment by deepeshgarg007** [comment]
@Mergifyio backport version-13-hotfix

**Pull request merged**